### PR TITLE
Contao 3.5.20

### PR DESCRIPTION
**Pull request by leofeyer**

### CODE REVIEW

**Review by discordier — APPROVED** [review]
LGTM :+1:

File: system/modules/core/library/Contao/Encryption.php [hunk 1]
```diff
@@ -93,9 +98,14 @@ public static function encrypt($varValue, $strKey=null)
 	 * @param string $strKey   An optional encryption key
 	 *
 	 * @return string The decrypted value
+	 *
+	 * @deprecated Deprecated since Contao 3.5, to be removed in Contao 5.
+	 *             Use the DCA load and save callbacks to implement your own encryption routines.
 	 */
 	public static function decrypt($varValue, $strKey=null)
 	{
+		@trigger_error('Using Encryption::decrypt() has been deprecated and will no longer work in Contao 5.0. Use the DCA load and save callbacks to implement your own encryption routines.', E_USER_DEPRECATED);
```
Comment (aschempp) [inline system/modules/core/library/Contao/Encryption.php]: This message is misleading, because the methods are not related to the DCA

**Review by aschempp — CHANGES_REQUESTED** [review]

File: system/modules/core/library/Contao/Encryption.php (hunk 1 above)
Comment (leofeyer) [inline system/modules/core/library/Contao/Encryption.php]: The load and save callbacks are not related to the DCA?

Comment (discordier) [inline system/modules/core/library/Contao/Encryption.php]: He is talking about the `\Encryption` class in general and it's methods.

Comment (leofeyer) [inline system/modules/core/library/Contao/Encryption.php]: Ah, I see. So what do you suggest then?

Comment (discordier) [inline system/modules/core/library/Contao/Encryption.php]: Why not simply omitting the mentioning of DCa at all?
"Using Encryption::decrypt() has been deprecated and will no longer work in Contao 5.0. Implement your own encryption routines."

We could also write a chapter in the `UPGRADING.md` dedicated to the overall issue and explaining the reasons of the deprecation and simply pass this link into the message.

Comment (aschempp) [inline system/modules/core/library/Contao/Encryption.php]: I'm fine with @discordier's message. Though as mentioned in the respective ticket, people won't know where the message is coming from…

**Comment by leofeyer** [comment]
All comments have been addressed. Please review again.

**Review by aschempp — APPROVED** [review]

**Review by ausi — APPROVED** [review]

File: system/modules/core/library/Contao/Encryption.php [hunk 2]
```diff
@@ -47,9 +47,14 @@ class Encryption
 	 * @param string $strKey   An optional encryption key
 	 *
 	 * @return string The encrypted value
+	 *
+	 * @deprecated Deprecated since Contao 3.5, to be removed in Contao 5.
```
Comment (Toflar) [inline system/modules/core/library/Contao/Encryption.php]: This message feels really wrong to me. Please **NEVER** implement your own encryption routine but use existing and tested libraries for this :-) The message as it is now implies you should invent something on your own :-D

File: system/modules/core/library/Contao/Widget.php [hunk 3]
```diff
@@ -1486,6 +1486,12 @@ public static function getAttributesFromDca($arrData, $strName, $varValue=null,
 			}
 		}
 
+		// Warn if someone uses the "encrypt" flag (see #8589)
+		if (isset($arrAttributes['encrypt']))
+		{
+			@trigger_error('Using the "encrypt" flag' . (!empty($strTable) && !empty($strField) ? ' on ' . $strTable . '.' . $strField : '') . ' has been deprecated and will no longer work in Contao 5.0. Use the load and save callbacks to implement your own encryption routine instead.', E_USER_DEPRECATED);
```
Comment (Toflar) [inline system/modules/core/library/Contao/Widget.php]: Same here, please adjust to something like `Use well tested third party libraries such as phpseclib`.

**Review by Toflar — APPROVED** [review]

**Comment by leofeyer** [comment]
The release has been scheduled for Monday, December 19th.

**Review by ausi — APPROVED** [review]

**Pull request merged**